File: python/translation_to_braille.py
from translator_mappings import (
    ENGLISH_TO_BRAILLE_MAPPINGS,
    NUMBERS_TO_BRAILLE_MAPPINGS,
    PUNCTUATION_TO_BRAILLE_MAPPINGS,
    CAPITAL_FOLLOWS_CASE,
    DECIMAL_FOLLOWS_CASE,
    NUMBER_FOLLOWS_CASE,
    SPACE_CASE
)

def translate_to_braille_char(char: str, previous_char: str) -> str:    
    if char == " ":
        return SPACE_CASE
    
    if previous_char.isnumeric():
        return NUMBERS_TO_BRAILLE_MAPPINGS[char]
    
    if char.isupper():
        return CAPITAL_FOLLOWS_CASE + ENGLISH_TO_BRAILLE_MAPPINGS[char.lower()]

    if char.isnumeric():
        return NUMBER_FOLLOWS_CASE + NUMBERS_TO_BRAILLE_MAPPINGS[char]
    
    if char in PUNCTUATION_TO_BRAILLE_MAPPINGS:
        return PUNCTUATION_TO_BRAILLE_MAPPINGS[char]
    
    return ENGLISH_TO_BRAILLE_MAPPINGS[char.lower()]

def translate_to_braille(text: str) -> str:
    translate_to_braille_array = []
    previous_char = ""

    for i, char in enumerate(text):
        previous_char = text[i - 1] if i > 0 else ""
        translate_to_braille_array.append(translate_to_braille_char(char, previous_char))
    
    return "".join(translate_to_braille_array)

File: javascript/translation-to-braille.js
import {
  ENGLISH_TO_BRAILLE_MAPPINGS,
  NUMBERS_TO_BRAILLE_MAPPINGS,
  PUNCTUATION_TO_BRAILLE_MAPPINGS,
  CAPITAL_FOLLOWS_CASE,
  DECIMAL_FOLLOWS_CASE,
  NUMBER_FOLLOWS_CASE,
  SPACE_CASE,
} from "./translator-mappings.js";

function isNumber(char) {
  return !isNaN(Number(char));
}

function translateToBrailleChar(char, previous_char) {
  if (char === " ") {
    return SPACE_CASE;
  }

  if (previous_char !== " " && isNumber(previous_char)) {
    return NUMBERS_TO_BRAILLE_MAPPINGS.get(char);
  }

  if (char === char.toUpperCase() && char !== char.toLowerCase()) {
    return CAPITAL_FOLLOWS_CASE + ENGLISH_TO_BRAILLE_MAPPINGS.get(char.toLowerCase());
  }

  if (char !== " " && isNumber(char)) {
    return NUMBER_FOLLOWS_CASE + NUMBERS_TO_BRAILLE_MAPPINGS.get(char);
  }

  if (PUNCTUATION_TO_BRAILLE_MAPPINGS.has(char)) {
    return PUNCTUATION_TO_BRAILLE_MAPPINGS.get(char);
  }

  return ENGLISH_TO_BRAILLE_MAPPINGS.get(char.toLowerCase());
}

export default function translateToBraille(text) {
  let translateToBrailleArray = [];
  let previous_char = " ";

  for (let i = 0; i < text.length; i++) {
    translateToBrailleArray.push(translateToBrailleChar(text[i], previous_char));
    previous_char = text[i];
  }

  return translateToBrailleArray.join("");
}
